### ISSUE TITLE
Review of topics

### ISSUE THREAD

**Issue by emitanaka** [numbats/learnr#6]
Following the development of topics in #2 , could you flag your reviewer for review of your topic (both the Moodle assessment + chapter):

- [x] Getting started with R and RStudio IDE (@jessica-wyleung)
- [x]  R Programming Basics (@DanyangDai)
- [x] How to ask for help (@mitchelloharawild)
- [x] Project structure (@mitchelloharawild)
- [x] Importing and exporting data (@jessica-wyleung)
- [x] Data wrangling with dplyr & tidyr (@mitchelloharawild)
- [x] Data visualisation (@jessica-wyleung)
- [x] Basic linear regression (@jessica-wyleung)
- [x] R Markdown basics (@emitanaka)

**Comment by DanyangDai** [comment]
Here are some comments for R Programming Basics @jessica-wyleung 

It is a very comperhensive chapter! Good job! Just some minor styling issues I picked up. 

- Vectors: 
> slide page 1. It seems that it the latex is not showing properly. But it would be that you want to show it that way. 
> slide page 6, script part. It stills shows $$ for your variables. Might need to look into the Latex problem for slides. 

- Matrices: 
> slide 1. Latex problem for the matrice display. 

- Check your understanding of matrices 
> Latex problem? 

- Data frame 
> In the exercise, it would be better to keep consistent assigning syntex with `<-`

- Conditional statement 
> slide 1: the last line is a bit cut off.

**Comment by jessica-wyleung** [comment]
@mitchelloharawild  Hi Mitch, I wonder if you can check if the latex for math is working properly on slides? I remember you fixed it before and I looked at the slides and it was working. So I am not sure what's wrong here? Thank you!

@DanyangDai  Thank you for the review! I have changed the `=` to `<-`. Sorry, Python habits sneaking up! I looked at the conditional statement slides and I can see the last line when I am doing full screen with my browser. I think it has something to do with the dimensions of the browser?

**Comment by DanyangDai** [comment]
@jessica-wyleung I just found a few other things for chapter 2. 

- for loop exercise, there is something wrong with the test function. It does not run successfully when running the solution. 
- while loops exercise, there are still a few `=` signs. 
- functions, exercise shows errors when running solutions.

**Comment by jessica-wyleung** [comment]
@DanyangDai  Thanks for the comment above! I fixed them.

I looked at Ch1 (Getting started with R) and Ch4 (import export data).

A few general things: 
(1) At the end of each chapter, you have a wonderful tab called "Additional Sources". Do you mean "Additional Resources" or is that like sources of your content?

(2) I think we should be consistent regarding the upper and lower case in the title of every tab of the chatpers. (At least within the same chapter it should be the same.) I notice that sometimes you have upper case D for Data in the title, while other words are lower case.

I understand some of these may have been fixed after George's comment but I will still put them here just in case.
Chapter 1.1 - the final page of the slides: image missing.
Chapter 1.2 - the third page of the slides: image missing.

Chapter 4.1 - slides page 1 - "will help you manage your datasets,(space is missing here)analyses and learning materials."
Chapter 4.1 - slides page 2 - "This is similar (to the) process **of** organizing..."

I don't have the perfect grammar - just trying to make the sentence sounds smoother.

**Comment by jessica-wyleung** [comment]
Hi @emitanaka

Chapter 6.8 -  Slides P.2 image missing.

Chapter 7.1 - I wonder if there are ways to make the font size more consistent between words and mathematical symbols. The   σ was so small on P. 6.

Chapter 7.3 - P. 3 font size a bit weird again

Chapter 7.7 model 4B - the latex didn't work.

**Comment by DanyangDai** [comment]
Hey @jessica-wyleung. Thank you so much for your detailed feed back. 

I've changed all the sources to resources. 

I have made all changes regarding to your suggestions for the gramma and lower case letters in the sub-headings. 

In terms of the missing images, I will try to fix this. @mitchelloharawild maybe you can help me with this?

**Comment by mitchelloharawild** [comment]
Image paths fixed for chapter1

**Comment by mitchelloharawild** [comment]
@DanyangDai 
## Chapter 3 review:
1. Not all mentions of 'error' needs to be red. Just have the first case in the paragraph be red, and mention that error messages often look scary, but they're just trying to help. I don't think you need to write your guesses for how the student feels, but instead focus on strategies on troubleshooting these errors. If they're feeling helpless, they already feel it and you don't need to say it. Same for if they're confused or overwhelmed. Instead say stuff like, don't panic!
2. I would spend more time explaining how to interpret error messages before you direct them to a search engine. Things like reading an error to find where the issue is would be helpful. Often there are hints, like object names in the errors, of even a `^` for syntax errors pointing out exactly where the issue is.
3. Sometimes there are no error messages, for instance if you run `mean(1:10` and forget the end bracket. Write about how the console shows `+` in this case (waiting for more code and the `)`) instead of `>`. Say that pressing `Esc` can get you back to `>` quickly.
4. Some debugging exercises would be useful here too. You should be able to provide code chunks in this section with small common errors for them to fix. Rather than using `<<...>>` for fill in the blank, just provide code with errors and use the testing section `???` to see if they've fixed the code.
5. Use `code` for code, and not for highlighting. you can bold, italicise or wrap in "" quotes if needed.
6. Run through a spell checker. I see things like "**Decrible** the problem"
7. 3.3 is a bit repetitive, you say to mention packages used but this is provided by `sessionInfo()`. Mention that simplifying your problem in producing a MRE can actually help you find out the issue yourself! A good resource for MREs is https://www.tidyverse.org/help/
8. Many of your **Don'ts** are handled by the reprex package or is repetitive. I would instead just have a bigger section on relative paths (possibly also linking to the projects chapter).
9. Spacing in guidance checklist is too large. I would add some extra text before the list too. If it's not exhaustive, then add some more bullet points. You can explain that this checklist cannot possibly cover all circumstances in the text below the title.
10. Given that this is for Monash, you can also add some course specific help options. This can be class specific, but often the Moodle forum or consultation times are a good place to ask for help. Also invite students to answer these types of questions to better practice the material.

**Comment by DanyangDai** [comment]
Hi @mitchelloharawild. Thank you for your detailed suggestions for chapter3. I have made justification accordingly. Please let me know if there is anything else I can do.

**Comment by mitchelloharawild** [comment]
@DanyangDai 
## Chapter 4 review
I think I'd prefer to have a separate chapter for data @emitanaka - it's a big topic and thematically different from project structure.

### Section 4.1
1. Some grammar issues in 4.1 slides, notably missing/inconsistent spaces and 'a R' not 'an R' in some places. Screenshots look okay, but their aspect ratio isn't right. Remove the width argument from your images.
2. Paths are written using `/` not `\`.
3. Include an example of an absolute path to illustrate why they're bad.
4. Show a screenshot of RStudio asking if the environment should be saved, making it clear which option they should choose and why for reproducibility.
5. Cool use of icons :+1:

### Section 4.2
1. `.` isn't a wildcard in R variable names. It's a wildcard in regular expressions when working with strings. However `.` is bad to use in R variable names due to the conventions of S3 methods (a topic we won't cover, but we can link to the chapter in advanced R).
2. `/data/propcessed` -> `/data/processed`, 'docutments' -> 'documents'. Run an en-GB spell checker.
3. Your suggested project structure seems pretty authoritative, and is not one that I follow. Others might prefer their `scripts` folder to be named `R` like is done in an R package, or `src` to be more language agnostic. `data` is pretty uncontroversial, but within it why not `cleaned` or `tidy` instead of `processed`? `results` are often stored in an `output` folder, or `reports`. What I'm getting at here is that there is no right answer nor standard approach, and so as long as the project is adequately organised and named for you and the people you work with it is okay. 
4. Why is this an exercise / not slides if there are no questions?

### Section 4.3
1. `.text` -> `.txt` file extension.
2. Missing `fwf`, another common format.
3. Missing databases, while not commonly used in uni courses it is needed in industry.

### Section 4.4 - 4.6
1. Paths starting with `/` are not relative. `/data/covid_cases_aus.csv` should be `data/covid_cases_aus.csv`.
2. I wouldn't bother introducing the base R functions for loading data, skip straight to the readr package to simplify section.
3. Preview the data to show how it is comma separated, and then read the data in.
4. Provide a reference to readr loading data faster than base.
5. Provide an example of reading an excel file
6. Provide some comparison between a tibble and a data.frame.
7. Merge 4.6 into the same chapter too. There's not much to learn here, just let students know that they can provide a URL to a data file too.

### Section 4.7
1. Use `readr` functions instead of base for writing data. Skip `.RData`, `.RDS` is better. Write about some advantages of saving data to `.RDS` too - it is a serialised data format which stores all information needed by R such as variable classes (what's a number vs date vs character string, etc.). However the big disadvantage is that it is R specific, and generally can't be read in by other languages).

**Comment by emitanaka** [comment]
@mitchelloharawild , I was expecting the topic of data (particularly tidy data) to be the subject of data wrangling

**Comment by mitchelloharawild** [comment]
@emitanaka The above was all for chapter 4 (import/export of data) for @DanyangDai.
I'm writing your chapter 5 review on wrangling and the tidyverse now.

**Comment by emitanaka** [comment]
If you spot small mistakes (e.g. typos), did you want to directly change it? I think that's more efficient and any big changes required can be the subject of discussion -- whether that be implemented this time or for next time.

**Comment by mitchelloharawild** [comment]
Yes.

**Comment by mitchelloharawild** [comment]
@emitanaka 
## Chapter 5 review

### Section 5.1
> dplyr is used for grammar of data manipulation. 

dplyr is *provides/implements* a grammar of data manipulation?

I actually don't think this is important for those learning R to know that it is a grammar. Perhaps mention more generally above that tidyverse packages aim to be cohesive with each other and provide simple building block functions for you to use in complex applications (a grammar).

Instead I think it is better to simply write that dplyr is a package that makes manipulating data easy. It helps with adding, renaming, and removing columns, computing new variables, joining datasets together and summarising your data.

Comparatively, your tidyr description is much more verbose. I think the details of one obs per row can be saved for the tidy section, and instead write more generally that it helps rearrange your data in a form that makes analysing your data easy.

Pipe subtopic is nice. I would write more about why pipes are relevant in the tidyverse. Maths typically isn't written in a sequence of operations like your pipe demo, and instead it is more familiar to write `exp(log(x) + 3)`.
Instead I wouldn't shy away from using simple tidyverse functions upfront with a real example that they can run, but have it very commented. Something like:

```r
# Read in data using readr
read_csv("market.csv") %>%
  # Create a new column called revenue using dplyr
  mutate(revenue = sales * price) %>%
  # Plot the data using ggplot2
  ggplot(aes(y = revenue, x = product)) + 
  geom_col()
```

Then compare that with the equivalent (but horrifying):
`ggplot(mutate(read_csv("market.csv"), revenue = sales * price), aes(y = revenue, x = product)) + geom_col()`

It's probably also worth mentioning the native pipe, both in terms of the widespread adoption of %>% in R having motivated |>, but also how it is faster but a comparatively a bit limited (no `.`, etc.). Justify why we use %>% in the materials.

### Section 5.2
Perhaps start with an overview slide of what dplyr can do for you? List key function groups including `select()`, `rename()`, `mutate()`, `summarise()`, joins, `filter()`, `arrange()`, groups?
I also think that this topic is too big for a single slide deck and quiz exercise.
Good examples and material, just lacks structure and flow of concepts. Broadly speaking I think separate sections can be written for dplyr's select/rename operations, row operations (filter, slice, arrange), transformation (mutate, group, summarise), and joins.
If you've put readers through the details of tidyselect, you may as well also introduce `across()` in the transformation topic.
When explaining joins, I would use two code examples. One without and one with the join to show how the results differ.
Join operations could benefit from the usual set visualisations.

### Section 5.3
The first code exercise could provide more structure to guide the function usage.
I think the first code exercise could come after some simpler questions, it seems like a single test of all of dplyr without giving feedback to misunderstandings of individual functions.

For problem 2, show the output of the datasets of size and colour.
Specify the `by = "product_name"` argument in the multi choice join operations. If you wanted to be tricky, have different key variable names and switch the order of `by = c("product" = "name")` with `by = c("name" = "product")`.

### Section 5.4

Why table1-3 and 5 but not 4?
I think you need more description for each table, both in their variable names and in the slide introducing them.
Write all datasets to the data/ folder so that they can be read in rather than created with `tibble()`. I don't expect new R users to create their own datasets with R that often, and the extra `read_csv()` reiteration should help with learning.
Don't replace NA with `999` - this is a dangerous practice that we don't need to demonstrate. Instead motivate an example with, suppose we found out that all missing values are actually `0` and not recorded correctly, lets `replace_na(list(return=0))` to fix this.

### Section 5.5

Again, show the data output before asking questions about it. Even though it's a `tribble()`, it adds another layer of complexity for translating code to imagined output to the question. Providing the code for this dataset may not even be needed here.

Your missing value question asks about something not covered in the learning materials. Also I wouldn't say 'best way' when there are alternatives like naniar. The best way is whatever is accurate and easiest/preferred by the analyst.

**Comment by emitanaka** [comment]
Chapter 5 is @jessica-wyleung . I'm good for Jessica to make the decisions for Ch 5.

**Comment by jessica-wyleung** [comment]
Thanks @mitchelloharawild  for the review. I have incorporated vast majority of them. 

I think identifying the key functions that the department needs while trying not to cover everything is important. I took the most common operations that I used with Python Pandas and tried to find the equivalent version in R when I was writing the content. I don't have much experience in which dplyr command is most commonly used. But I am happy to add them in the next iteration as you suggested.
